### DISCUSSION TITLE
RPI-PICO-I2C-LCD lib slow work(Solved)

### DISCUSSION THREAD

**Discussion by oldschoolmountainbike** [micropython/micropython#16219] [RP2040 / Pico] [answered]
Hello! 
After adding another new function code started work slowly. I simulated this problem - I send a couple of words one by one to the 20x4 display, everything is fine, but if I add 60 empty functions that are not even called, the text starts to display very slowly.

Are there any solutions to add more function without slowing down the code?

Raspberry Pi Pico  with RP2040
MicroPython v1.23.0  
IDE:Thonny
```py
 from machine import Pin, I2C, Timer
import time
from i2c_lcd import I2cLcd

 

I2C_ADDR_LCD = 0x27  # lcd I2C 2004
i2c = I2C(0, scl=Pin(21), sda=Pin(20), freq=400000)  
lcd = I2cLcd(i2c, I2C_ADDR_LCD, 4, 20)   
#########################################################################

def test1():
    i=0
    return i

# ..... for test1 to test60

def test60():
    i=0
    return i
############################################################
rows, cols = 4, 20

 
lcd_buffer = [" " * cols for _ in range(rows)]  

def update_lcd(new_data):
 
    global lcd_buffer

    for i in range(rows):
         
        if not isinstance(new_data[i], str):
            new_data[i] = str(new_data[i])   

 
        if len(new_data[i]) < cols:
            new_data[i] += " " * (cols - len(new_data[i]))
 
        elif len(new_data[i]) > cols:
            new_data[i] = new_data[i][:cols]

 
    for row in range(rows):
        for col in range(cols):
            if lcd_buffer[row][col] != new_data[row][col]:
 
                lcd.move_to(col, row)
                lcd.putstr(new_data[row][col])
        
 
        lcd_buffer[row] = new_data[row]

new_lcd_data = [
                            'some text1', #  
                            'some text2 ', #  
                            'some text3 ', #  
                            'some text4 '] #  
            
new_lcd_data1 = [
                            'text test1',  
                            'text test2',  
                            'text test3',  
                            'text test4']  
 
while True:
    update_lcd(new_lcd_data)
    time.sleep(0.1)
    update_lcd(new_lcd_data1)
    time.sleep(0.1)
```

**Comment by peterhinch** [comment]
Please format your code as [described here](https://github.com/orgs/micropython/discussions/9111).

**Reply by oldschoolmountainbike** [reply]
sorry, done.

**Comment by oldschoolmountainbike** [comment]
For test I put global variables inside a function and nothing change with slowly work,only wrong work :) But when i tried send result not to Lcd,as here
```py
                lcd.move_to(col, row)
                lcd.putstr(new_data[row][col])
```
I delete this two function and add 'print' for send results to Thommy, all working fast. These two function doesn't like a lot of empty functions, but why and how it fix ?

**Comment by shariltumin** [comment]
This is strange. Having many and unused functions does not affect the execution of a function. Maybe you can try using different I2C?

on my board:

```
MPY: soft reboot
MicroPython v1.25.0-preview.4.g9ea8d2a03.dirty on 2024-11-13; WaveShare RP2040 Zero with RP2040
>>>
Remount local directory ./ at /remote
>>> import machine
>>> i2c = machine.I2C(0)
>>> i2c
I2C(0, freq=399361, scl=5, sda=4, timeout=50000)
>>> 
```

**Reply by oldschoolmountainbike** [reply]
> This is strange. Having many and unused functions does not affect the execution of a function. Maybe you can try using different I2C?
> 
> on my board:
> 
> ```
> MPY: soft reboot
> MicroPython v1.25.0-preview.4.g9ea8d2a03.dirty on 2024-11-13; WaveShare RP2040 Zero with RP2040
> >>>
> Remount local directory ./ at /remote
> >>> import machine
> >>> i2c = machine.I2C(0)
> >>> i2c
> I2C(0, freq=399361, scl=5, sda=4, timeout=50000)
> >>> 
> ```

thanks for the answer! I tried change ports to scl=5, sda=4 ,reconnect Lcd pins and it didn't work, 
```py
  File "<stdin>", line 11, in <module>
  File "i2c_lcd.py", line 22, in __init__
OSError: [Errno 5] EIO
```
Maybe problem with i2c_lcd libs, can you share your lib ,please?

**Reply by shariltumin** [reply]
I don't have an LCD display to try. My board is running v1.24 (v1.25.0-preview.4). You are running v1.23.0. You can try using machine.I2C(1). Just wondering, is your display a 5V or 3V3 device? But I can't understand why function definitions in the script affect I2C speed. 

You might want to try: https://github.com/T-622/RPI-PICO-I2C-LCD/tree/main

**Reply by oldschoolmountainbike** [reply]
Thanks for this link! I tried it-no changes. 
Lcd is 5V, and bus is 5v too, I know about the risks and will add a level converter soon.
"But I can't understand why function definitions in the script affect I2C speed."  - I notice- If I put Lcd function first and empty function after  calling lcd  -all working good.

**Comment by oldschoolmountainbike** [comment]
And if put Lcd function first and empty function after  calling lcd -all work good too

**Comment by mendenm** [comment]
Is it possible that you are causing hash table collisions for function name lookups with very similar function names?  I haven't looked at how MP does hashing.  Try 60 randomly generated function names (use a python program to generate the functions, then paste them into this code).  See if it runs fast then.

**Reply by oldschoolmountainbike** [reply]
Thanks for the answer! This problem occurred in the code that controls the audio processor, in this code the functions had different names, but the problem was the same - adding any function greater than a certain number, even empty and which is not called and the display slows down. To make it easier to show the problem, I simulated a simple code. But if I put Lcd function first and empty function after calling lcd -all work good

**Reply by shariltumin** [reply]
As Jos said, you may be running out of memory. Putting all the extra (non-active) functions after everything does not affect the execution of the lcd function. Is it possible to do ```print(gc.mem_free())```, before, during and after the lcd function?

**Reply by oldschoolmountainbike** [reply]
I agree with your opinion.  Thanks,I did it. Here are the results

```py
start 218672
inside Lcd function,begin 216656
inside  "for in range"  216432
inside  "for in range"  216480
inside  "for in range"  216528
inside  "for in range"  216576
inside Lcd function,end 216576
point before while 216704
end 216512
```
same mem test without 60 empty function
```py
start 222544
inside function,begin 222080
inside  "for in"  221856
inside  "for in"  221904
inside  "for in"  221952
inside  "for in"  222000
inside function for,end 222000
end before while 222160
end 221936
```

last two points here-
```py
print('end before while',gc.mem_free()) 
while True:
    update_lcd(new_lcd_data)
    print('one')
    time.sleep(0.1)
    update_lcd(new_lcd_data1)
    print('two')
    time.sleep(0.1)
    print('end',gc.mem_free()) 
```

**Reply by shariltumin** [reply]
I don't see any memory issues here. You can try to measure the time it takes to complete update_lcd().

```
>>> import time
>>> now = time.ticks_us
>>> gap = time.ticks_diff
>>> n = now();time.sleep_ms(5);p = gap(now(),n); print(p)
5109
```

so you can try these:

```py
print('end before while',gc.mem_free())
while True:
    n = now(); update_lcd(new_lcd_data); print(gap(now(),n))
    print('one')
    time.sleep(0.1)
    n = now(); update_lcd(new_lcd_data1); print(gap(now(),n))
    print('two')
    time.sleep(0.1)
    print('end',gc.mem_free())
```

**Reply by oldschoolmountainbike** [reply]
> I don't see any memory issues here. You can try to measure the time it takes to complete update_lcd().

Thanks for this code! 
Rusult with empty functions are
```py
1301140
1299473
```
Without -
```py
378326
378878
```
 Inside lcd function I have interesting result 
```py
    n = now() ####### inside point

    for row in range(rows):
        for col in range(cols):
            if lcd_buffer[row][col] != new_data[row][col]:

                p = gap(now(),n); print('inside ',p)

                n = now() ###### Lcd time point

                lcd.move_to(col, row)
                lcd.putstr(new_data[row][col])

                p = gap(now(),n); print('Lcd time ',p)
```
```py
inside  97
inside  18673
inside  37255
inside  55771
....
inside  1312743

Lcd time  18141
```
without empty function-

```py
inside  91
inside  5825
inside  11628
inside  17381
....
inside  403187

Lcd time  5326
```
Empty function part - 2076
```py
n = now()
def test1():
    i=0
    return i

# ..... for test1 to test60

def test60():
    i=0
    return i
p = gap(now(),n); print('empty ',p)
```

each time the running time of lcd.move_to,lcd.putstr is added to the cycle and if there are empty functions, this time increases four times. (from 5 to almost 20) . Interesting, how empty functions did it

**Reply by shariltumin** [reply]
I don't know what is happening here. Please do the same timing test, but this time disable the ```lcd.move_to()``` and ```lcd.putstr()``` and replace them with ```print()```. Do tests with and without the empty functions.

Please use a bidirectional level shifter before you continue with the lcd tests.I don't think the RP2040's GPIO pins are 5V tolerant.

I'm out of ideas here, but you can try
1. Use the newest firmware
2. Use a different i2c_lcd lib
3. Try with a different MCU, for example ESP32

**Reply by oldschoolmountainbike** [reply]
thanks for the recommendations! I disable the ```py lcd.move_to()``` and ```py lcd.putstr()``` and replace them with ```py print()``` result is ``` 13494```, without empty function ``` 11290 ```, 2000 of difference  , same result was with  test only empty functions-
```py 
n = now()
def test1():
    i=0
    return i
......................
def test60():
    i=0
    return i
p = gap(now(),n); print(p)
```
to exclude or confirm the problem with "i2c lcd" lib, I removed these librarie and ```py lcd.move_to()``` ,and ```py lcd.putstr()``` from the code,. Added eeprom 24c64 to the i2c bus and made reading 10 bytes
```py
    for row in range(rows):
        for col in range(cols):
            if lcd_buffer[row][col] != new_data[row][col]:
 
                start_address = 0x0000  ###here was lcd.move_to()
                read_data = read_from_eeprom(start_address, 10) ###here was lcd.putstr()
```
And result without empty function - 39219 , with - 40100. Empty functions have no effect here!
The code has a similar structure and logic, but it is not affected by empty functions. The problem is with the display library.   Thanks for the link above, I have exactly these libraries and I will try to ask the author a question
```py
from machine import Pin, I2C
.....
i2c = I2C(0, scl=Pin(21), sda=Pin(20), freq=400000)
......
i2c.writeto(EEPROM_ADDR, address_bytes)
read_data = i2c.readfrom(EEPROM_ADDR, num_bytes)
```

**Reply by oldschoolmountainbike** [reply]
For fast work in i2c_lcd lib change "utime" to "time" and delete a lot of gc.collect().
[https://github.com/T-622/RPI-PICO-I2C-LCD/issues/17#issue-2092503824](url)

**Reply by oldschoolmountainbike** [reply]
Thank you very much for your help!

**Comment by Josverl** [comment]
do you pre-compile your code, or is all the code in a single large file that 
- needs to be loaded from flash into RAM ( block 1 ) 
- then compiled 
- then run, allocating blocks of memory after  (block 1) 
- also I think the line `lcd_buffer = [" " * cols for _ in range(rows)]  ` is fragmenting memory more than it should.

And thus fragmenting memory .

In my experience i've been able to run large and complex applications in surprisingly little memory , once I learned how to avoid fragmenting it. 
and to be fair - that is a bit of an art as it is not simple to look at/assess the memory consumption of running code.

**Reply by oldschoolmountainbike** [reply]
Thank for the answer! Yes , all as you wrote, I don't pre-compile code.  This is my first project and I am not familiar with precompilation yet, please tell me how best to do this?

**Comment by oldschoolmountainbike** [comment, marked as answer]
SOLVED!
For fast work in i2c_lcd lib change "utime" to "time" and delete a lot of gc.collect().
[https://github.com/T-622/RPI-PICO-I2C-LCD/issues/17#issue-2092503824](url)
Thanks all for helping!
